include/aws/crt/Optional.h

```diff
@@ -86,18 +94,56 @@ namespace Aws
 
                 if (m_value)
                 {
-                    m_value->~T();
+                    if (other.m_value)
+                    {
+                        *m_value = *other.m_value;
+                    }
+                    else
+                    {
+                        m_value->~T();
+                        m_value = nullptr;
+                    }
+
+                    return *this;
                 }
 
                 if (other.m_value)
                 {
-                    new (&m_storage) T(*other.m_value);
+                    new (&m_storage) T();
                     m_value = reinterpret_cast<T *>(&m_storage);
-                    other.m_value = nullptr;
+                    *m_value = *other.m_value;
                 }
-                else
+
+                return *this;
+            }
+
+            template <typename U = T> Optional<T> &operator=(const Optional<U> &other)
+            {
+                if (this == &other)
                 {
-                    m_value = nullptr;
+                    return *this;
+                }
+
+                if (m_value)
+                {
+                    if (other.m_value)
+                    {
+                        *m_value = *other.m_value;
+                    }
+                    else
+                    {
+                        m_value->~T();
+                        m_value = nullptr;
+                    }
+
+                    return *this;
+                }
+
+                if (other.m_value)
+                {
+                    new (&m_storage) T();
+                    m_value = reinterpret_cast<T *>(&m_storage);
+                    *m_value = *other.m_value;
                 }
 
                 return *this;
```

Comment:
This comment is for the `operator=(Optional<U> &&other)` lower down:

It should look a lot like the other operator=(Optional:
```
if (this == &other)
{
    return *this;
}

if (m_value)
{
    if (other.m_value)
    {
        *m_value = std::forward(*other.m_value);
    }
    else
    {
        m_value->~T();
        m_value = nullptr;
    }

    return *this;
}

if (other.m_value)
{
    new (&m_storage) T(std::forward(*other.m_value));
    m_value = reinterpret_cast<T *>(&m_storage);
}

return *this;
```
```